### PR TITLE
Speed up threshold_local function by fixing call to _supported_float_type

### PR DESCRIPTION
**Pull request by kevinmeetooa**
Fix call to _supported_float_type(dtype) in threshold_local() function.

We currently pass the image itself instead of its dtype to the _supported_float_type function, forcing it to iterate over the entire image to get its dtype.

Passing the image's dtype directly leads to a much faster execution time of the threshold_local() function

In my case, for (2480, 3508) images, the execution time goes down from 20s on average to 2s on average with this fix.

### CODE REVIEW

**Comment by lagru** [comment]
Great catch @kevinmeetooa! I'm happy to take the fix as is, though I think it would be good to make sure that such an error doesn't happen in the future. 

I think the way would be to add something like 
```python
try:
    input_dtype = input_dtype.dtype
except AttributeError:
    ...
```
before https://github.com/scikit-image/scikit-image/blob/c90bfccbc930262eebd1149b10660bfa4f56ec61/skimage/_shared/utils.py#L729.

What do you think?

File: skimage/_shared/utils.py
Comment (lagru) [inline skimage/_shared/utils.py:733]: The test's fail with `TypeError: Cannot interpret '<attribute 'dtype' of 'numpy.generic' objects>' as a data type`. I think this should do
```suggestion
    if hasattr(input_dtype, 'dtype'):
        input_dtype = input_dtype.dtype
    elif isinstance(input_dtype, Iterable) and not isinstance(input_dtype, str):
        return np.result_type(*(_supported_float_type(d) for d in input_dtype))
    else:
        input_dtype = np.dtype(input_dtype)
```

Comment (kevinmeetooa) [reply skimage/_shared/utils.py]: This caused more issues (AttributeError: 'getset_descriptor' object has no attribute 'char')
I used a try/except instead. 

I try casting `input_dtype` with `input_dtype = np.dtype(input_dtype)` and I then catch two exceptions:

- `TypeError` for cases when an array is passed as argument (`TypeError: Cannot construct a dtype from an array`). 
- ValueError when an array with two different float values

Comment (kevinmeetooa) [reply skimage/_shared/utils.py]: This caused more issues (AttributeError: 'getset_descriptor' object has no attribute 'char')
I used a try/except instead. 

I try casting `input_dtype` with `input_dtype = np.dtype(input_dtype)`. This works in cases where input_dtype is a numpy dtype. I then catch two exceptions:

- `TypeError` for cases when an array is passed as argument (`TypeError: Cannot construct a dtype from an array`). This is then handled in the `Except` block with `getattr(input_dtype, 'dtype', input_dtype)`

- ValueError for cases where `input_dtype` is an array with two different float values. I had this issue in the `test_supported_float_dtype_sequence` test where _supported_float_type((np.float16, np.float32)) is tested

The tests look fine now

File: skimage/_shared/utils.py
Comment (lagru) [inline skimage/_shared/utils.py:734]: Oh, seems indeed a bit more complicated. Thanks for iterating on this. While this seems to work I'm not happy with the logic flow here as it's starting to be hard to reason about it. E.g. I think `isinstance(input_dtype, str)` is obsolete with the current solution but I'm not sure if there is truly no case where it is reached.

Rather than trying to handle all kinds of inputs we should probably be more restrictive. How about
```python
    if isinstance(input_dtype, tuple):
        return np.result_type(*(_supported_float_type(d) for d in input_dtype))
    input_dtype = np.dtype(input_dtype)
```
which just makes sure to fail explicitly for everything except dtypes or tuple or dtypes. That avoids iteration on arrays and also makes testing for `str` unnecessary. This would probably require updating the docstring a bit and some tests. Do you agree?

Comment (kevinmeetooa) [reply skimage/_shared/utils.py:729]: Done, this is much cleaner indeed. 
With this, I also found that the `restoration.uft.ir2tf` function also called `_supported_float_type(imp_resp)` instead of `_supported_float_type(imp_resp.dtype)`, so I've updated this function as well.

Comment (lagru) [reply skimage/_shared/utils.py:729]: Perfect! Could you update the docstring as well? :pray: Except for that looks ready to merge.

Comment (kevinmeetooa) [reply skimage/_shared/utils.py:729]: done :slightly_smiling_face:

**Review by lagru — APPROVED** [review]
Well done and thank you your patience and for what I'm assuming to be your first contribution to scikit-image. :D

FYI, in scikit-image we have the [policy](https://scikit-image.org/docs/dev/skips/1-governance.html#decision-making-process) to require at least core developers to check the code for non-trivial PRs. So once another maintainer has approved, it'll be merged.

**Comment by kevinmeetooa** [comment]
alright, thank you!

**Review by jarrodmillman — APPROVED** [review]

**Comment by kevinmeetooa** [comment]
> > We currently pass the image itself instead of its dtype to the _supported_float_type function, forcing it to iterate over the entire image to get its dtype.
> 
> I don't think this is accurate. The image (array) is simply a pointer, so it does not matter much whether you send the image itself or its dtype?

I'm not quite sure I understand. From my understanding, it does matter since sending the image itself causes you to enter the `if` condition at [ln. 729](https://github.com/scikit-image/scikit-image/blob/main/skimage/_shared/utils.py#L729). 
Modifying the [function call](https://github.com/scikit-image/scikit-image/blob/main/skimage/filters/thresholding.py#L230) to pass `image.dtype` instead of `image` prevents from entering the `if` condition above, avoiding an iteration on the entire image.

**Comment by stefanv** [comment]
> > > We currently pass the image itself instead of its dtype to the _supported_float_type function, forcing it to iterate over the entire image to get its dtype.
> > 
> > 
> > I don't think this is accurate. The image (array) is simply a pointer, so it does not matter much whether you send the image itself or its dtype?
> 
> I'm not quite sure I understand. From my understanding, it does matter since sending the image itself causes you to enter the `if` condition at [ln. 729](https://github.com/scikit-image/scikit-image/blob/main/skimage/_shared/utils.py#L729). Modifying the [function call](https://github.com/scikit-image/scikit-image/blob/main/skimage/filters/thresholding.py#L230) to pass `image.dtype` instead of `image` prevents from entering the `if` condition above, avoiding an iteration on the entire image.

I see what you're saying; the function was incorrectly invoked. If this were a public facing function, we'd try and handle the arguments more robustly, but since it's internal this is perfectly fine. Thanks!

**Comment by yuval-ngtnuma** [comment]
Perhaps there should be benchmarks on these functions?

**Pull request merged**
